### PR TITLE
[TECH-debt] `Tasking::get_quotations` test coverage fix

### PR DESCRIPTION
**Pull request by andher1802**
See changelog.

Checklist:
* [x] Test coverage
* [x] Version bump
* [x] Changelog update

### CODE REVIEW

File: up42/tasking.py [hunk 1]
```diff
@@ -130,26 +133,29 @@ def get_quotations(
...
-        return self._query_paginated_output(url)
+        sort_order = ",desc" if descending else ",asc"
+        query_params = {
+            key: str(value)
+            for key, value in {
+                "workspaceId": workspace_id,
+                "id": quotation_id,
+                "orderId": order_id,
+                "decision": set(decision) if decision else None,
+                "sort": sortby + sort_order,
```
Comment (javidq) [inline up42/tasking.py:144]: `"sort": utils.SortingField(sortby, descending)`

File: up42/tasking.py [hunk 2]
```diff
@@ -130,26 +133,29 @@ def get_quotations(
...
+            }.items()
+            if value
+        }
+        quotations = []
+        current_page = 0
+        while True:
```
Comment (javidq) [inline up42/tasking.py:150]: Let's reuse the approach from Order module

File: up42/tasking.py [hunk 3]
```diff
@@ -130,26 +133,29 @@ def get_quotations(
...
-                for single_decision in decision:
-                    url += f"&decision={single_decision}"
-            else:
-                logger.warning(
-                    "decision values are NOT_DECIDED, ACCEPTED, REJECTED, otherwise decision filter values ignored."
-                )
-        return self._query_paginated_output(url)
+        sort_order = ",desc" if descending else ",asc"
+        query_params = {
+            key: str(value)
```
Comment (javidq) [inline up42/tasking.py:138]: Why do we need str here?

File: up42/tasking.py [hunk 4]
```diff
@@ -130,26 +133,29 @@ def get_quotations(
...
-        url = host.endpoint(f"/v2/tasking/quotation?page=0&sort={sort}")
-        if quotation_id is not None:
-            url += f"&id={quotation_id}"
-        if workspace_id is not None:
-            url += f"&workspaceId={workspace_id}"
-        if order_id is not None:
-            url += f"&orderId={order_id}"
-        if decision is not None:
-            decisions_validation = (
-                single_decision in ["NOT_DECIDED", "ACCEPTED", "REJECTED"] for single_decision in decision
```
Comment (javidq) [inline up42/tasking.py:143]: Let's add these as type hints to decision

File: tests/test_tasking.py [hunk 5]
```diff
@@ -34,6 +36,8 @@
         ),
     ),
 }
+QUOTATION_ID = str("some-quotation-id")
```
Comment (javidq) [inline tests/test_tasking.py:39]: Why do you need str?

File: tests/test_tasking.py [hunk 6]
```diff
@@ -34,6 +36,8 @@
...
+ORDER_ID = str("some-order-id")
```
Comment (javidq) [inline tests/test_tasking.py:40]: We have a constant for it

File: tests/test_tasking.py [hunk 7]
```diff
@@ -110,16 +114,56 @@ def test_should_construct_order_parameters(
...
+        if workspace_id:
+            query_params["workspaceId"] = workspace_id
+        if order_id:
+            query_params["orderId"] = order_id
+        if decision:
+            query_params["decision"] = set(decision)
+        if sort_by:
+            query_params["sort"] = sort_by + sort_order
+
+        for page in range(total_pages):
```
Comment (javidq) [inline tests/test_tasking.py:148]: Let's simplify this to two pages with inline mocking

File: tests/test_tasking.py [hunk 8]
```diff
@@ -110,16 +114,56 @@ def test_should_construct_order_parameters(
...
+            url = f"{constants.API_HOST}/v2/tasking/quotation"
+            query_params["page"] = page
+            query = urllib.parse.urlencode(query_params)
+            url += query and f"?{query}"
+            response = {
+                "content": [{"id": str(uuid.uuid4())} for _ in range(10)],
+                "totalPages": total_pages,
+            }
+            requests_mock.get(url=url, json=response)
+        get_quotations = tasking_obj.get_quotations(
```
Comment (javidq) [inline tests/test_tasking.py:158]: get_quotations -> quotations

File: tests/test_tasking.py [hunk 9]
```diff
@@ -110,16 +114,56 @@ def test_should_construct_order_parameters(
...
+            quotation_id=quotation_id,
+            workspace_id=workspace_id,
+            order_id=order_id,
+            decision=decision,
+            sortby=sort_by,
+            descending=descending,
+        )
+        assert len(get_quotations) == 10 * total_pages
```
Comment (javidq) [inline tests/test_tasking.py:166]: We need to check its content

File: tests/test_tasking.py [hunk 10]
```diff
@@ -110,16 +114,56 @@ def test_should_construct_order_parameters(
...
+        if sort_by:
+            query_params["sort"] = sort_by + sort_order
+
+        for page in range(total_pages):
+            url = f"{constants.API_HOST}/v2/tasking/quotation"
+            query_params["page"] = page
+            query = urllib.parse.urlencode(query_params)
+            url += query and f"?{query}"
+            response = {
+                "content": [{"id": str(uuid.uuid4())} for _ in range(10)],
```
Comment (javidq) [inline tests/test_tasking.py:154]: Let's switch to "id": f"id{idx}" to be able to check the content

File: tests/test_tasking.py [hunk 11]
```diff
@@ -110,16 +114,56 @@ def test_should_construct_order_parameters(
...
+        quotation_id: Optional[str],
+        workspace_id: Optional[str],
+        order_id: Optional[str],
+        decision: Optional[List[str]],
+        sort_by: str,
+        descending: bool,
+    ):
+        total_pages = 4
+        query_params: dict[str, Any] = {}
+        sort_order = ",desc" if descending else ",asc"
```
Comment (javidq) [inline tests/test_tasking.py:136]: let's add this to parameterization

File: tests/test_tasking.py [hunk 12]
```diff
@@ -110,16 +114,56 @@ def test_should_construct_order_parameters(
...
+        if quotation_id:
+            query_params["id"] = quotation_id
+        if workspace_id:
+            query_params["workspaceId"] = workspace_id
+        if order_id:
+            query_params["orderId"] = order_id
+        if decision:
+            query_params["decision"] = set(decision)
```
Comment (javidq) [inline tests/test_tasking.py:144]: set is not needed

File: tests/test_tasking.py [hunk 13]
```diff
@@ -110,16 +114,56 @@ def test_should_construct_order_parameters(
...
-    assert all(quotation["workspaceId"] == constants.WORKSPACE_ID for quotation in get_workspace_quotations)
-    get_accepted_quotations = tasking_mock.get_quotations(decision=["ACCEPTED"])
-    assert all(quotation["decision"] == "ACCEPTED" for quotation in get_accepted_quotations)
-    get_decided_quotations = tasking_mock.get_quotations(decision=["ACCEPTED", "REJECTED"])
-    assert all(quotation["decision"] in ["ACCEPTED", "REJECTED"] for quotation in get_decided_quotations)
+    @pytest.mark.parametrize("quotation_id", [None, QUOTATION_ID])
+    @pytest.mark.parametrize("workspace_id", [None, constants.WORKSPACE_ID])
+    @pytest.mark.parametrize("order_id", [None, ORDER_ID])
+    @pytest.mark.parametrize("decision", [None, ["NOT_DECIDED", "ACCEPTED", "REJECTED"]])
+    @pytest.mark.parametrize("sort_by", ["updatedAt"])
```
Comment (javidq) [inline tests/test_tasking.py:121]: there's no point in parameterizing this

**Review by javidq — CHANGES_REQUESTED** [review]

File: tests/test_tasking.py (hunk 11 above)
Comment (andher1802) [reply tests/test_tasking.py:136]: I have parametrize descending and set the value according to it. Is that what you meant?

File: up42/tasking.py (hunk 3 above)
Comment (andher1802) [reply up42/tasking.py:134]: No, actually it is an error due to `decision` being an array

Comment (andher1802) [reply up42/tasking.py]: No, actually it was an error as desicion is an array.

File: up42/tasking.py [hunk 14]
```diff
@@ -14,6 +14,8 @@
 
 Geometry = Union[catalog.Geometry, geom.Point]
 
+QuotationStatuses = Literal["NOT_DECIDED", "ACCEPTED", "REJECTED"]
```
Comment (javidq) [inline up42/tasking.py:17]: Statuses -> Status

File: up42/tasking.py [hunk 15]
```diff
@@ -130,26 +132,28 @@ def get_quotations(
...
+
+        def get_pages(endpoint: str):
+            while True:
+                response = self.session.get(host.endpoint(endpoint), params=params).json()
+                yield response["content"]
+                params["page"] += 1
+                if params["page"] == response["totalPages"]:
+                    break
+
+        return [quotation_info for page in get_pages("/v2/tasking/quotation") for quotation_info in page]
```
Comment (javidq) [inline up42/tasking.py:156]: quotation_info -> quotation

File: tests/test_tasking.py [hunk 16]
```diff
@@ -110,16 +110,61 @@ def test_should_construct_order_parameters(
...
+        expected = [{"id": f"id{idx}"} for idx in [1, 2, 3, 4]]
+
+        quotations = tasking_obj.get_quotations(
+            quotation_id=quotation_id,
+            workspace_id=workspace_id,
+            order_id=order_id,
+            decision=decision,
+            descending=descending,
+        )
+        assert len(quotations) == 4
```
Comment (javidq) [inline tests/test_tasking.py:166]: Not needed

File: tests/test_tasking.py [hunk 17]
```diff
@@ -110,16 +110,61 @@ def test_should_construct_order_parameters(
...
+            query_params["id"] = quotation_id
+        if workspace_id:
+            query_params["workspaceId"] = workspace_id
+        if order_id:
+            query_params["orderId"] = order_id
+        if decision:
+            query_params["decision"] = decision
+        base_url = f"{constants.API_HOST}/v2/tasking/quotation"
+
+        query = urllib.parse.urlencode(query_params, doseq=True, safe="")
```
Comment (javidq) [inline tests/test_tasking.py:139]: Let's simplify mocking to 
```python
        for page in [0, 1]:
            query_params["page"] = page
            query = urllib.parse.urlencode(query_params, doseq=True, safe="")
            url = base_url + (query and f"?{query}")
            offset = page * 2
            response = {
                "content": expected[offset : offset + 2],
                "totalPages": 2,
            }
            requests_mock.get(url=url, json=response)
```
and we can also remove setting page on L128

**Review by javidq — CHANGES_REQUESTED** [review]

**Review by javidq — APPROVED** [review]

**Pull request merged**